### DISCUSSION TITLE
Integration with rollback networking libs?

### DISCUSSION THREAD

**Discussion by JRMurr** [jcornaz/heron#168] [Ideas]
Im working on a multiplayer game and would like to use a rollback lib like https://github.com/gschup/bevy_ggrs, https://github.com/HouraiTeahouse/backroll-rs/, or maybe https://github.com/ErnWong/crystalorb (this one's bevy integration is not a nice so less likely to be used). 

I may be misunderstanding how these libs work since im mostly in the research phase still but the issue for ggrs at least is that it has its own schedule where all game play logic needs to be registered. Since the physics for heron are run in its own stage(s) physics would not be run correctly when rolling back and re simulating frames. 

There are potentially other issues with how physics steps vs frames are handled between the two libs.

My main question for this post is if anyone has any examples of heron being used with one of these libs. If not any idea for what changes i would need to make in heron to allow more customization on when heron logic is run?

**Comment by JRMurr** [comment]
If this is something that is out of scope of heron i totally understand. Would really love to keep using heron since its so simple but if this kind of customization reduces the simplicity I understand and can work with rapier directly in my game.

Based on the ggrs dev rapier might not be a good fit for rollback due to how long de-serializing can take https://discord.com/channels/691052431525675048/747940465936040017/926212530181931008

**Reply by jcornaz** [reply]
Hi, 

I don't know ggrs, and will probably not have time to look at it seriously anytime soon.

As for heron to support it I don't know, it depends what would be needed in heron.

If it is a question to let the user choose in what stage the heron systems must run, that'd be doable. We could provide a variant of the phyics plugin that take a the stages as argument:

```rust
fn main() {
  App::build()
    .add_plugins(DefaultPlugins)
    .add_plugin(StagedPhysicsPlugin {
        pre_physics_stage: CoreStage::PreUpdate,
        physics_stage: CoreStage::Update,
    })
    // ...
    .run();
}
```

And then heron would run the physics systems in that stages (that should be already registered before installing the physics plugin).

Would that be enough?

**Reply by JRMurr** [reply]
Thanks for the response!

Yea i think something like that would help greatly and probably get me almost there. I may also need labels on the systems to make the logic order more deterministic but i can handle that if i run into it.

The other thing is for ggrs at least is you need to tell it what components/resources need to be loaded/saved when a rollback occurs. https://github.com/gschup/bevy_ggrs/blob/12ba7a8d4355a5db445d28ab65714b71d4decdd9/examples/box_game/box_game_p2p.rs#L64

I think for something like that it should hopefully just be enough for the heron types to be public.

**Reply by jcornaz** [reply]
> The other thing is for ggrs at least is you need to tell it what components/resources need to be loaded/saved when a rollback occurs. 

That can be done by user code right? Or would heron need to do something specific about it?

**Reply by JRMurr** [reply]
Yea all user code, when I work with it I'll pr an example to show all the things that need to be registered to ggrs

**Reply by jcornaz** [reply]
I created #170. I am not sure when I'll find time to implement it though. But it shouldn't be too hard in case you are interested to contribute ;-)

**Reply by JRMurr** [reply]
Yea i would love to help! Ill try it sometime this week
